File: web-frontend/src/services/aiService.js
import axios from 'axios';

const OPENROUTER_API_KEY = import.meta.env.VITE_OPENROUTER_API_KEY;
const SITE_URL = 'http://localhost:5173'; // Update for production
const SITE_NAME = 'Carbon Sleuth';

const PRIMARY_MODEL = import.meta.env.VITE_AI_MODEL || "google/gemini-2.0-flash-lite-preview-02-05:free";
const FALLBACK_MODEL = "google/gemma-2-9b-it:free"; // Robust fallback

const callOpenRouter = async (model, contextData, userQuery) => {
    const systemPrompt = `
    You are an industrial AI assistant for a chemical plant monitoring dashboard.
    
    CONTEXT DATA:
    ${JSON.stringify(contextData, null, 2)}

    USER QUERY: "${userQuery}"

    INSTRUCTIONS:
    1.  **PRIORITY:** If the user's request corresponds to a dashboard action (filtering, searching, navigating), **output the ACTION TAG ALONE**. Do not write a paragraph unless necessary.
    2.  **Filtering/Searching:** If the user asks to "Show me X", "Filter by X", "List X", "Show components with X" or simply types a keyword like "Pressure":
        -   **Step 1:** Generate a clean, comprehensive **Markdown Filtered Table** of the matching equipment from the context data. Include all relevant columns.
        -   **Step 2:** Append the Action Tag: \`|ACTION:SEARCH:X|\` at the very end.
    3.  **Navigation:** If the user asks for "Analytics", "Charts", "Dashboard", or "Home":
        -   Response: \`|ACTION:NAVIGATE:/analytics|\` (or other path)
    4.  **Analysis:** Only if the user asks a complex question ("Why is pressure high?", "Explain root cause"), then provide a text answer.
    5.  **Tables:** If you must generate a table, use standard Markdown tables.
    
    EXAMPLES:
    - User: "Show me warning components" -> AI: \`|ACTION:SEARCH:warning|\`
    - User: "Go to analytics" -> AI: \`|ACTION:NAVIGATE:/analytics|\`
    - User: "Pressure" -> AI: \`|ACTION:SEARCH:pressure|\`
    `;

    return await axios.post("https://openrouter.ai/api/v1/chat/completions", {
        "model": model,
        "messages": [
            { "role": "system", "content": systemPrompt },
            { "role": "user", "content": userQuery }
        ]
    }, {
        headers: {
            "Authorization": `Bearer ${OPENROUTER_API_KEY}`,
            "HTTP-Referer": window.location.origin,
            "X-Title": SITE_NAME,
            "Content-Type": "application/json"
        }
    });
};

export const queryAI = async (contextData, userQuery) => {
    if (!OPENROUTER_API_KEY) {
        return "Error: OpenRouter API Key is missing. Please configure VITE_OPENROUTER_API_KEY in .env";
    }

    try {
        console.log(`Attempting AI Query with Primary Model: ${PRIMARY_MODEL}`);
        const response = await callOpenRouter(PRIMARY_MODEL, contextData, userQuery);
        return response.data.choices[0].message.content;

    } catch (primaryError) {
        console.warn(`Primary Model (${PRIMARY_MODEL}) Failed. Switching to Fallback (${FALLBACK_MODEL}).`, primaryError.message);

        try {
            const fallbackResponse = await callOpenRouter(FALLBACK_MODEL, contextData, userQuery);
            return fallbackResponse.data.choices[0].message.content + "\n\n*(Answered by Fallback Model)*";
        } catch (fallbackError) {
            console.error("All AI Models Failed Details:", {
                primary: primaryError.message,
                fallback: fallbackError.message
            });
            return "Sorry, both primary and fallback AI models are currently unavailable. Please check your API usage or internet connection.";
        }
    }
};

File: desktop-frontend/app/ai_service.py
import os
import requests
import json

# Configuration
SITE_NAME = "Carbon Sleuth"
OPENROUTER_API_KEY = os.getenv("VITE_OPENROUTER_API_KEY") or "sk-or-v1-..." # Fallback or load from .env if possible
AI_MODEL = os.getenv("VITE_AI_MODEL") or "google/gemma-2-9b-it:free"

def generate_ai_response(user_query: str, context_data: dict) -> dict:
    """
    Generates an AI response based on the user query and dashboard context.
    Mirroring the logic from web-frontend/src/services/aiService.js.
    
    This function:
    1. Constructs a system prompt with the current data context.
    2. Sends a request to the LLM provider (OpenRouter).
    3. Parses the response for 'Action Tags' (e.g., |ACTION:SEARCH:X|).
    
    :param user_query: The question or command from the user.
    :param context_data: JSON object containing current dashboard statistics/data.
    :return: Dict containing 'response' (text) and 'action' (parsed command).
    """
    if not user_query:
        return None

    # Construct System Prompt
    system_prompt = f"""
    You are Carbon Sleuth AI, an expert industrial analyst.
    Current Site: {SITE_NAME}
    
    EXPLICIT INSTRUCTION FOR "SHOW ME" / FILTERING REQUESTS:
    If the user asks to "Show me", "List", "Filter", or "Find" items (e.g., "Show me warning items", "List high pressure pumps"):
    1. FIRST, generate a Markdown Table containing the relevant data items.
    2. THEN, after the table, append the action tag: |ACTION:SEARCH:<filter_term>|
    
    DO NOT just say "Here are the items". YOU MUST DISPLAY THE TABLE.
    
    CONTEXT DATA:
    {json.dumps(context_data, indent=2)}
    """

    headers = {
        "Authorization": f"Bearer {OPENROUTER_API_KEY}",
        "Content-Type": "application/json",
        "HTTP-Referer": "http://localhost:3000", # Dummy for desktop
        "X-Title": SITE_NAME,
    }

    payload = {
        "model": AI_MODEL,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_query}
        ],
        "temperature": 0.7,
        "max_tokens": 1000
    }

    try:
        response = requests.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers=headers,
            json=payload,
            timeout=15
        )
        response.raise_for_status()
        data = response.json()
        
        content = data['choices'][0]['message']['content']
        
        # Parse Action Tags (Simple parser)
        action = None
        if "|ACTION:SEARCH:" in content:
            parts = content.split("|ACTION:SEARCH:")
            term = parts[1].split("|")[0].strip()
            action = {"type": "SEARCH", "payload": term}
            # Remove tag from display
            content = content.replace(f"|ACTION:SEARCH:{term}|", "")
            
        return {"response": content, "action": action}

    except Exception as e:
        print(f"AI Service Error: {e}")
        import traceback
        traceback.print_exc()
        return {"response": f"I encountered an error connecting to the AI service: {str(e)}", "action": None}
